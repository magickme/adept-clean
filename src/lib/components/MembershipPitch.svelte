<script>
	let { membershipInfo, logoUrl, linkUrl, isDarkMode } = $props();
</script>

<div class="bg-kohl p-8">
	<div class="bg-kohl/80 text-ivory p-4 rounded-lg shadow-md max-w-3xl mx-auto ring-1 ring-ivory/10">
		<div class="flex items-center space-x-4">
			<a href={linkUrl} aria-label="Membership link">
				<div class="flex-shrink-0">
					<img src={logoUrl} alt="Logo" class="h-8 w-8 text-cerulean" width="32" height="32" />
				</div>
			</a>
			<a href={linkUrl} aria-label="Membership link">
				<div class="flex-1 text-sm font-sourceSans">
					<p>{membershipInfo}</p>
				</div>
			</a>
		</div>
	</div>
</div>
