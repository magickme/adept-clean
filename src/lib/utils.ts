type ClassValue = string | number | boolean | undefined | null | { [key: string]: boolean };

export function cn(...inputs: ClassValue[]): string {
  return inputs.filter(Boolean).join(' ');
} 