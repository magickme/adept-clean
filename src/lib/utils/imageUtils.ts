export function getOptimizedAvatarUrl(url: string, size: number = 40): string {
  if (!url) return '';
  
  // Handle pravatar.cc URLs
  if (url.includes('pravatar.cc')) {
    // Extract the user parameter
    const userMatch = url.match(/\?u=([^&]+)/);
    const user = userMatch ? userMatch[1] : '';
    
    // Return optimized URL with correct size
    return `https://i.pravatar.cc/${size}?u=${user}`;
  }
  
  return url;
}

export function getAvatarImageProps(size: number = 40) {
  return {
    width: size,
    height: size,
    loading: 'lazy' as const,
    decoding: 'async' as const,
    fetchpriority: 'low' as const,
  };
}

export function getFeaturedImageProps(isAboveFold: boolean = false) {
  return {
    loading: isAboveFold ? 'eager' as const : 'lazy' as const,
    decoding: 'async' as const,
    fetchpriority: isAboveFold ? 'high' as const : 'low' as const,
    width: 800,
    height: 600,
  };
}

export function getOptimizedImageUrl(url: string, width: number): string {
  if (!url) return '';
  
  // Handle assets.adept.magick.me URLs
  if (url.includes('assets.adept.magick.me')) {
    const filename = url.split('/').pop();
    const ext = filename?.split('.').pop();
    const name = filename?.replace(`.${ext}`, '');
    
    // Return optimized URL with correct size and WebP format
    return `https://assets.adept.magick.me/${name}-${width}.webp`;
  }
  
  return url;
} 