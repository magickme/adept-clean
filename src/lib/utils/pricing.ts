export function formatPrice(price: number): string {
  return (price / 100).toFixed(0);
}

export function getCheckoutUrl(courseId: string, plan: { id: string; price: number }): string {
  return `/checkout?courseId=${courseId}&planId=${plan.id}&price=${plan.price}`;
} 