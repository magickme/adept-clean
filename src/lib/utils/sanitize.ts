import DOMPurify from 'isomorphic-dompurify';

export function sanitizeHtml(html?: string): string {
  return DOMPurify.sanitize(html || '');
} 