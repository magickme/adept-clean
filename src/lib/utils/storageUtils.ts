/**
 * Request persistent storage using modern navigator.storage API
 * @returns Promise<boolean> - Whether persistent storage was granted
 */
export async function requestPersistentStorage(): Promise<boolean> {
  try {
    // Check if the Storage API is supported
    if (!navigator?.storage?.persist) {
      console.warn('Storage API not supported');
      return false;
    }

    // Check if persistence is already granted
    const isPersisted = await navigator.storage.persisted();
    if (isPersisted) {
      return true;
    }

    // Request persistence
    const granted = await navigator.storage.persist();
    return granted;
  } catch (error) {
    console.error('Error requesting persistent storage:', error);
    return false;
  }
}

/**
 * Get storage usage and quota information
 * @returns Promise<{usage: number, quota: number}> - Storage usage info in bytes
 */
export async function getStorageInfo(): Promise<{ usage: number; quota: number }> {
  try {
    if (!navigator?.storage?.estimate) {
      return { usage: 0, quota: 0 };
    }

    const estimate = await navigator.storage.estimate();
    return {
      usage: estimate.usage || 0,
      quota: estimate.quota || 0
    };
  } catch (error) {
    console.error('Error getting storage info:', error);
    return { usage: 0, quota: 0 };
  }
} 