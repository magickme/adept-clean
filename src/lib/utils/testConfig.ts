export interface TestConfig {
  readonly timeouts: {
    readonly render: number;
    readonly animation: number;
    readonly network: number;
    readonly interaction: number;
  };
  readonly limits: {
    readonly maxTextLength: number;
    readonly maxArrayLength: number;
    readonly maxNestedDepth: number;
  };
  readonly features: {
    readonly useRealTimers: boolean;
    readonly mockNetwork: boolean;
    readonly isolateComponents: boolean;
  };
  readonly environment: {
    readonly isCi: boolean;
    readonly isDebug: boolean;
    readonly testEnv: 'unit' | 'integration' | 'e2e';
  };
}

const DEFAULT_CONFIG: TestConfig = {
  timeouts: {
    render: 1000,
    animation: 300,
    network: 5000,
    interaction: 100
  },
  limits: {
    maxTextLength: 1000,
    maxArrayLength: 100,
    maxNestedDepth: 3
  },
  features: {
    useRealTimers: false,
    mockNetwork: true,
    isolateComponents: true
  },
  environment: {
    isCi: process.env.CI === 'true',
    isDebug: process.env.DEBUG === 'true',
    testEnv: (process.env.TEST_ENV || 'unit') as TestConfig['environment']['testEnv']
  }
};

function deepFreeze<T extends object>(obj: T): Readonly<T> {
  Object.keys(obj).forEach(prop => {
    const value = obj[prop as keyof T];
    if (value && typeof value === 'object') {
      deepFreeze(value as object);
    }
  });
  return Object.freeze(obj);
}

export const getTestConfig = (overrides?: Partial<TestConfig>): Readonly<TestConfig> => {
  const baseConfig = structuredClone(DEFAULT_CONFIG);
  
  if (overrides) {
    Object.entries(overrides).forEach(([key, value]) => {
      if (value && typeof value === 'object') {
        Object.assign(baseConfig[key as keyof TestConfig], value);
      } else {
        (baseConfig as any)[key] = value;
      }
    });
  }
  
  return deepFreeze(baseConfig);
};

const frozenConfig = deepFreeze(structuredClone(DEFAULT_CONFIG));
export const testConfig = frozenConfig; 