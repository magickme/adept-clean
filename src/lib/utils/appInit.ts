import { requestPersistentStorage, getStorageInfo } from './storageUtils';

/**
 * Initialize app features and permissions
 */
export async function initializeApp() {
  // Request persistent storage
  const persistentStorage = await requestPersistentStorage();
  if (persistentStorage) {
    // Get storage info for monitoring
    const { usage, quota } = await getStorageInfo();
    console.debug('Storage info:', {
      usage: Math.round(usage / 1024 / 1024) + 'MB',
      quota: Math.round(quota / 1024 / 1024) + 'MB',
      percentage: Math.round((usage / quota) * 100) + '%'
    });
  }
} 