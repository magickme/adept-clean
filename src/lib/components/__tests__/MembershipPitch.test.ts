import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/svelte/svelte5';
import MembershipPitch from '../MembershipPitch.svelte';
import { membershipPitchTestUtils as utils } from './utils/membershipPitchTestUtils';

describe('MembershipPitch', () => {
  describe('Basic Rendering', () => {
    it('renders with all props', () => {
      const props = utils.createProps();
      render(MembershipPitch, { props });

      expect(utils.queries.getMembershipInfo()).toBeInTheDocument();
      expect(utils.queries.getLogo()).toHaveAttribute('src', props.logoUrl);
      expect(utils.queries.getLinks()).toHaveLength(2);
    });

    it('matches snapshot', () => {
      const { container } = render(MembershipPitch, { props: utils.createProps() });
      expect(container).toMatchSnapshot();
    });

    it('applies correct base styles', () => {
      const { container } = render(MembershipPitch, { props: utils.createProps() });
      utils.verify.classes(container.querySelector(utils.selectors.outer), utils.classes.outer);
      utils.verify.classes(container.querySelector(utils.selectors.inner), utils.classes.inner);
    });
  });

  describe('Props Validation', () => {
    it('handles empty membership info', () => {
      render(MembershipPitch, { props: utils.createProps({ membershipInfo: '' }) });
      const textLink = utils.queries.getTextLink();
      expect(textLink).toBeInTheDocument();
      expect(textLink?.querySelector('p')?.textContent?.trim()).toBe('');
    });

    it('handles missing logo URL', () => {
      render(MembershipPitch, { props: utils.createProps({ logoUrl: undefined }) });
      expect(utils.queries.getLogo().getAttribute('src')).toBeFalsy();
    });

    it('handles invalid logo URL', () => {
      const onError = vi.fn();
      render(MembershipPitch, { props: utils.createProps({ logoUrl: 'invalid-url' }) });
      
      const logo = utils.queries.getLogo();
      logo.addEventListener('error', onError);
      logo.dispatchEvent(new Event('error'));
      
      expect(onError).toHaveBeenCalled();
    });

    it('handles special characters in membership info', () => {
      const specialText = '<script>alert("xss")</script>Special & <b>Bold</b>';
      render(MembershipPitch, { 
        props: utils.createProps({ membershipInfo: specialText }) 
      });
      const textContent = utils.queries.getTextLink()?.textContent;
      expect(textContent).toBe(specialText);
      expect(utils.queries.getTextLink()?.innerHTML).not.toBe(specialText);
    });

    it('handles extremely long membership info', () => {
      const longText = 'a'.repeat(1000);
      render(MembershipPitch, { 
        props: utils.createProps({ membershipInfo: longText })
      });
      expect(utils.queries.getTextLink()).toBeInTheDocument();
      utils.verify.layout(document.body);
    });

    it('handles extremely long URLs', () => {
      const longUrl = `https://example.com/${'a'.repeat(500)}`;
      render(MembershipPitch, { 
        props: utils.createProps({ linkUrl: longUrl })
      });
      utils.queries.getLinks().forEach(link => {
        expect(link).toHaveAttribute('href', longUrl);
      });
    });
  });

  describe('Accessibility and Layout', () => {
    it('has proper ARIA labels and alt text', () => {
      render(MembershipPitch, { props: utils.createProps() });
      utils.queries.getLinks().forEach(utils.verify.linkAttributes);
      expect(utils.queries.getLogo()).toHaveAttribute('alt', 'Logo');
    });

    it('maintains responsive layout structure', () => {
      const { container } = render(MembershipPitch, { props: utils.createProps() });
      utils.verify.layout(container);
    });
  });

  describe('Dark Mode', () => {
    it('applies correct dark mode styles', () => {
      const { container } = render(MembershipPitch, { 
        props: utils.createProps({ isDarkMode: true }) 
      });
      utils.verify.darkMode(container);
    });

    it('maintains layout in dark mode', () => {
      const { container } = render(MembershipPitch, { 
        props: utils.createProps({ isDarkMode: true }) 
      });
      utils.verify.layout(container);
    });

    it('preserves link and logo visibility in dark mode', () => {
      render(MembershipPitch, { props: utils.createProps({ isDarkMode: true }) });
      expect(utils.queries.getLogoLink()).toBeVisible();
      expect(utils.queries.getTextLink()).toBeVisible();
    });
  });

  describe('Performance', () => {
    it('renders efficiently with large content', () => {
      const start = performance.now();
      const longText = 'a'.repeat(10000);
      render(MembershipPitch, { 
        props: utils.createProps({ membershipInfo: longText })
      });
      const end = performance.now();
      expect(end - start).toBeLessThan(100); // 100ms threshold
    });

    it('handles rapid re-renders', async () => {
      const { rerender } = render(MembershipPitch);
      
      // Batch rerenders or reduce count
      const updates = Array(10).fill().map((_, i) => ({ 
        someProps: `value${i}` 
      }));

      for (const props of updates) {
        rerender(props);
        // Add small delay if needed
        await new Promise(r => setTimeout(r, 10)); 
      }
    });
  });

  describe('Accessibility', () => {
    it('meets color contrast requirements', () => {
      const { container } = render(MembershipPitch, { props: utils.createProps() });
      utils.verify.contrast(container);
    });

    it('preserves tab order', () => {
      render(MembershipPitch, { props: utils.createProps() });
      const tabbableElements = utils.queries.getTabbableElements();
      expect(tabbableElements.length).toBe(2); // Two links
      expect(tabbableElements[0]).toBe(utils.queries.getLogoLink());
      expect(tabbableElements[1]).toBe(utils.queries.getTextLink());
    });
  });
}); 