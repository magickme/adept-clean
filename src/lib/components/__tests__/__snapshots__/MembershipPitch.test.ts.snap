// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MembershipPitch > Basic Rendering > matches snapshot 1`] = `
<div>
  <div
    class="bg-kohl p-8"
  >
    <div
      class="bg-gray-800 text-white p-4 rounded-lg shadow-md max-w-3xl mx-auto"
    >
      <div
        class="flex items-center space-x-4"
      >
        <a
          aria-label="Membership link"
          href="https://example.com"
        >
          <div
            class="flex-shrink-0"
          >
            <img
              alt="Logo"
              class="h-8 w-8 text-blue-400"
              src="/test-logo.png"
            />
          </div>
        </a>
         
        <a
          aria-label="Membership link"
          href="https://example.com"
        >
          <div
            class="flex-1 text-sm"
          >
            <p>
              Test membership info
            </p>
          </div>
        </a>
      </div>
    </div>
  </div>
  
</div>
`;
